### PR TITLE
Issue #10466 review session documentation.

### LINKED ISSUE #10466: Review HTTP session documentation
**Jetty version(s)**
12

**Description**
The HTTP session documentation in the programming guide needs a review.

* Drop references to eeN or Servlets, everything should be explained in terms of core only
* Verify that the class diagrams are still up to date
* Verify that configuration of the various elements is still valid (properties, setter methods, etc.)
* Review the prose in light of Jetty 12.

### PR DESCRIPTION
**Pull request by janbartel**
Closes #10466

Also fix session config context init param names and add missing code to configure SessionHandlers via context init params, and added test for config.

### CODE REVIEW

File: documentation/jetty-documentation/src/main/asciidoc/programming-guide/server/sessions/session-architecture.adoc [hunk 1]
```diff
@@ -13,18 +13,31 @@
 
 [[pg-server-session-architecture]]
 ==== Session Architecture
+==== Introduction
```
Comment (sbordet) [inline documentation/jetty-documentation/src/main/asciidoc/programming-guide/server/sessions/session-architecture.adoc]: I'd remove this line.

FYI the number of `=` is the depth of the section, so if you mean `Introduction` to be a child of `Session Architecture`, it should have 5 `=`.

Same goes for other section titles below.

File: documentation/jetty-documentation/src/main/asciidoc/programming-guide/server/sessions/session-architecture.adoc [hunk 2]
```diff
@@ -13,18 +13,31 @@
...
+Jetty session support has been architected to provide a core implementation that is independent of the servlet specification.
```
Comment (sbordet) [inline documentation/jetty-documentation/src/main/asciidoc/programming-guide/server/sessions/session-architecture.adoc]: `Servlet` always has a capital `S`.

File: documentation/jetty-documentation/src/main/asciidoc/programming-guide/server/sessions/session-architecture.adoc [hunk 3]
```diff
@@ -13,18 +13,31 @@
...
+This allows programmers who use core Jetty - without the servlet api - to still have classic servlet session-like support for their `Requests` and `Handlers`.
```
Comment (sbordet) [inline documentation/jetty-documentation/src/main/asciidoc/programming-guide/server/sessions/session-architecture.adoc]: `api` -> `APIs`

`Requests` -> ``Request``s as the class does not have a final `s`. Same for `Handlers` and blow `Servlets`, `Filters`, etc.

File: documentation/jetty-documentation/src/main/asciidoc/programming-guide/server/sessions/session-architecture.adoc [hunk 4]
```diff
@@ -13,18 +13,31 @@
...
+Jetty also adapts these core classes to each of the various servlet specification environments to deliver classic `HttpSessions` for `Servlets`, `Filters` etc.
+
+We provide full support for the session lifecycle, in addition to L1 and L2 caching, and a number of options for persisting session data.
+These features are all available to both core and servlet environments.
+
+==== Terminology
+
+In order to set the scene for our discussion of sessions in Jetty, here are some of the most important concepts that we  will refer to throughout the documentation.
+
+SessionIdManager::: responsible for allocation of unique session ids. A `Server` can have at most 1 `SessionIdManager`.
```
Comment (sbordet) [inline documentation/jetty-documentation/src/main/asciidoc/programming-guide/server/sessions/session-architecture.adoc]: Description lists only have to `:`, so this line should be:

`SessionIdManager`::

It's typical to have them in their own line, and the description in a new line below it.

File: documentation/jetty-documentation/src/main/asciidoc/programming-guide/server/sessions/session-architecture.adoc [hunk 5]
```diff
@@ -13,18 +13,31 @@
...
-SessionCache::: is an L1 cache of in-use `Session` objects
-Session::: is a stateful object representing a `HttpSession`
+Jetty also adapts these core classes to each of the various servlet specification environments to deliver classic `HttpSessions` for `Servlets`, `Filters` etc.
+
+We provide full support for the session lifecycle, in addition to L1 and L2 caching, and a number of options for persisting session data.
+These features are all available to both core and servlet environments.
+
+==== Terminology
+
+In order to set the scene for our discussion of sessions in Jetty, here are some of the most important concepts that we  will refer to throughout the documentation.
```
Comment (sbordet) [inline documentation/jetty-documentation/src/main/asciidoc/programming-guide/server/sessions/session-architecture.adoc]: Never use "we", either "you" or passive.
This whole sentence could go, it's colloquial and adds nothing.

File: documentation/jetty-documentation/src/main/asciidoc/programming-guide/server/sessions/session-architecture.adoc [hunk 6]
```diff
@@ -13,18 +13,31 @@
...
 
-Terminology::
-SessionIdManager::: is responsible for allocation of session ids
-HouseKeeper:::  is responsible for orchestrating the detection and removal of expired sessions
-SessionHandler::: is responsible for managing the lifecycle of sessions within its associated context
-SessionCache::: is an L1 cache of in-use `Session` objects
-Session::: is a stateful object representing a `HttpSession`
+Jetty also adapts these core classes to each of the various servlet specification environments to deliver classic `HttpSessions` for `Servlets`, `Filters` etc.
+
+We provide full support for the session lifecycle, in addition to L1 and L2 caching, and a number of options for persisting session data.
```
Comment (sbordet) [inline documentation/jetty-documentation/src/main/asciidoc/programming-guide/server/sessions/session-architecture.adoc]: We -> Jetty (don't use first person plural).

File: documentation/jetty-documentation/src/main/asciidoc/programming-guide/server/sessions/session-sessioncache.adoc [hunk 7]
```diff
@@ -14,30 +14,30 @@
...
-It then looks for a `SessionDataStoreFactory` bean on the server to produce a `SessionDataStore` instance to use with the `SessionCache`.
+. allow the `SessionManager` to create one lazily at startup.
+The `SessionManager` looks for a `SessionCacheFactory` bean on the `Server` to produce the `SessionCache` instance.
+It then looks for a `SessionDataStoreFactory` bean on the `Server` to produce a `SessionDataStore` instance to use with the `SessionCache`.
 
-. pass a fully configured `SessionCache` instance to the `SessionHandler`.
+. pass a fully configured `SessionCache` instance to the `SessionManager`.
 You are responsible for configuring both the `SessionCache` instance and its `SessionDataStore` 
 
 More on ``SessionDataStore``s xref:pg-server-session-datastore[later], in this section we will concentrate on the `SessionCache` and `SessionCacheFactory`.
```
Comment (sbordet) [inline documentation/jetty-documentation/src/main/asciidoc/programming-guide/server/sessions/session-sessioncache.adoc]: Avoid "we".

Something like:

You can find more information on `SessionDataStore` xref:...[here], while this section details `SessionCache` and...

File: documentation/jetty-documentation/src/main/asciidoc/programming-guide/server/sessions/session-sessioncache.adoc [hunk 8]
```diff
@@ -14,30 +14,30 @@
...
 
 
-The link:{javadoc-url}/org/eclipse/jetty/server/session/AbstractSessionCache.html[AbstractSessionCache] provides most of the behaviour of ``SessionCache``s.
+The link:{javadoc-url}/org/eclipse/jetty/session/AbstractSessionCache.html[AbstractSessionCache] provides most of the behaviour of ``SessionCache``s.
 If you are implementing a custom `SessionCache` we strongly recommend you extend this base class, as the Servlet Specification has many subtleties and extending the base class ensures that your implementation will take account of them.
```
Comment (sbordet) [inline documentation/jetty-documentation/src/main/asciidoc/programming-guide/server/sessions/session-sessioncache.adoc]:
```suggestion
If you are implementing a custom `SessionCache` you should extend this base class, as the Servlet specification has many subtleties, and extending the base class ensures that your implementation will account for them.
```

File: documentation/jetty-documentation/src/main/asciidoc/programming-guide/server/sessions/session-sessioncache.adoc [hunk 9]
```diff
@@ -68,61 +68,61 @@ To prevent his, enable this feature and the `SessionCache` will ensure that if a
...
 
-The link:{javadoc-url}/org/eclipse/jetty/server/session/DefaultSessionCache.html[DefaultSessionCache] retains `Session` objects in memory in a `ConcurrentHashMap`.
+The link:{javadoc-url}/org/eclipse/jetty/session/DefaultSessionCache.html[DefaultSessionCache] retains `ManagedSession` objects in memory in a `ConcurrentHashMap`.
 It is suitable for non-clustered and clustered deployments.
 For clustered deployments, a sticky load balancer is *strongly* recommended, otherwise you risk indeterminate session state as the session bounces around multiple nodes.
 
-It implements the link:{javadoc-url}/org/eclipse/jetty/server/session/SessionCache.html#shutdown()[SessionCache.shutdown()] method.
+It implements the link:{javadoc-url}/org/eclipse/jetty/session/SessionCache.html#shutdown()[SessionCache.shutdown()] method.
 
 It also provides some statistics on sessions, which are convenient to access either directly in code or remotely via jmx:
```
Comment (sbordet) [inline documentation/jetty-documentation/src/main/asciidoc/programming-guide/server/sessions/session-sessioncache.adoc]: jmx -> JMX

File: documentation/jetty-documentation/src/main/asciidoc/programming-guide/server/sessions/session-sessioncache.adoc [hunk 10]
```diff
@@ -133,11 +133,11 @@ include::../../{doc_code}/org/eclipse/jetty/docs/programming/server/session/Sess
 [[pg-server-session-customcache]]
 ===== Implementing a Custom SessionCache
 
-As previously mentioned, we highly recommend that you extend the link:{javadoc-url}/org/eclipse/jetty/server/session/AbstractSessionCache.html[AbstractSessionCache].
+As previously mentioned, we highly recommend that you extend the link:{javadoc-url}/org/eclipse/jetty/session/AbstractSessionCache.html[AbstractSessionCache].
```
Comment (sbordet) [inline documentation/jetty-documentation/src/main/asciidoc/programming-guide/server/sessions/session-sessioncache.adoc]:
```suggestion
As previously mentioned, you should extend the link:{javadoc-url}/org/eclipse/jetty/session/AbstractSessionCache.html[AbstractSessionCache].
```

File: documentation/jetty-documentation/src/main/asciidoc/programming-guide/server/sessions/session-sessionhandler.adoc [hunk 11]
```diff
@@ -14,115 +14,113 @@
...
+There are `SessionHandlers` for both the core and the various servlet environments.
+
+Note that in the servlet environments, each `ServletContextHandler` or `WebAppContext` has at most a single `SessionHandler`.
 
 ===== Configuration
 
-The majority of configuration for the link:{javadoc-url}/org/eclipse/jetty/server/session/SessionHandler.html[SessionHandler] can be done via `web.xml` `<session-config>` declarations, or the `javax.servlet.SessionCookieConfig` api.
-There are also a few jetty-specific configuration options that we will cover here:
+The majority of configuration for the `SessionHandler` can be done via `web.xml` `<session-config>` declarations, or the `javax.servlet.SessionCookieConfig` api.
+There are also a few jetty-specific configuration methods that we will cover here:
```
Comment (sbordet) [inline documentation/jetty-documentation/src/main/asciidoc/programming-guide/server/sessions/session-sessionhandler.adoc]:
```suggestion
There are also a few Jetty-specific configuration methods:
```

File: documentation/jetty-documentation/src/main/java/org/eclipse/jetty/docs/programming/server/session/SessionDocs.java [hunk 12]
```diff
@@ -72,14 +75,44 @@ public void defaultSessionIdManagerWithHouseKeeper()
...
+            //end::coresession[]
+        }
+        catch (Exception e)
+        {
+            e.printStackTrace();
+        }
+    }
     
-    public void servletContextWithSessionHandler()
+    public void ee10ServletContextWithSessionHandler()
```
Comment (sbordet) [inline documentation/jetty-documentation/src/main/java/org/eclipse/jetty/docs/programming/server/session/SessionDocs.java]: We only document the latest EE version, so ee10 here is implicit and you can drop it.

File: documentation/jetty-documentation/src/main/java/org/eclipse/jetty/docs/programming/server/session/SessionDocs.java [hunk 13]
```diff
@@ -72,14 +75,44 @@ public void defaultSessionIdManagerWithHouseKeeper()
...
     {
         //tag:schsession[]
         Server server = new Server();
         
-        ServletContextHandler context = new ServletContextHandler("/foo", ServletContextHandler.SESSIONS);
-        SessionHandler sessions = context.getSessionHandler();
+        org.eclipse.jetty.ee10.servlet.ServletContextHandler context = new org.eclipse.jetty.ee10.servlet.ServletContextHandler("/foo", org.eclipse.jetty.ee10.servlet.ServletContextHandler.SESSIONS);
```
Comment (sbordet) [inline documentation/jetty-documentation/src/main/java/org/eclipse/jetty/docs/programming/server/session/SessionDocs.java]: Should not be necessary to use the full qualified class name here, as we should not document `ee9` and does not clash with a core class.

File: documentation/jetty-documentation/src/main/java/org/eclipse/jetty/docs/programming/server/session/SessionDocs.java [hunk 14]
```diff
@@ -72,14 +75,44 @@ public void defaultSessionIdManagerWithHouseKeeper()
...
+        org.eclipse.jetty.ee10.servlet.SessionHandler sessions = context.getSessionHandler();
```
Comment (sbordet) [inline documentation/jetty-documentation/src/main/java/org/eclipse/jetty/docs/programming/server/session/SessionDocs.java]: If this class clashes with the same core class, perhaps use `var sessions = context.getSessionHandler()` -- it's shorter and more to the point.

File: jetty-core/jetty-deploy/src/test/resources/etc/webdefault.xml [hunk 15]
```diff
@@ -52,15 +52,15 @@
...
       <param-value>127.0.0.1</param-value> 
     </context-param> 
     <context-param>
-      <param-name>org.eclipse.jetty.ee9.servlet.SessionPath</param-name>
+      <param-name>org.eclipse.jetty.servlet.SessionPath</param-name>
       <param-value>/</param-value>
     </context-param>
     <context-param>
-      <param-name>org.eclipse.jetty.ee9.servlet.MaxAge</param-name>
+      <param-name>org.eclipse.jetty.servlet.MaxAge</param-name>
```
Comment (sbordet) [inline jetty-core/jetty-deploy/src/test/resources/etc/webdefault.xml]: Are these parameters really missing the `eeN` in the package name?

File: jetty-core/jetty-session/src/main/java/org/eclipse/jetty/session/SessionConfig.java [hunk 16]
```diff
@@ -20,6 +20,46 @@
 // TODO the managed attribute descriptions need review
 public interface SessionConfig
 {
+    /**
+     * Session cookie name.
+     * Defaults to <code>JSESSIONID</code>, but can be set with the
+     * <code>org.eclipse.jetty.session.SessionCookie</code> context init parameter.
+     */
+    String __SessionCookieProperty = "org.eclipse.jetty.session.SessionCookie";
```
Comment (sbordet) [inline jetty-core/jetty-session/src/main/java/org/eclipse/jetty/session/SessionConfig.java:28]: We dropped double underscores for statics long time ago.
Use all upper case.

File: jetty-core/jetty-session/src/main/java/org/eclipse/jetty/session/SessionConfig.java [hunk 17]
```diff
@@ -20,6 +20,46 @@
...
+     */
+    String __SessionDomainProperty = "org.eclipse.jetty.session.SessionDomain";
+    String __DefaultSessionDomain = null;
+    /**
+     * Session Path.
+     * If this property is set as a ServletContext InitParam, then it is
+     * used as the path for the session cookie.  If it is not set, then
+     * the context path is used as the path for the cookie.
+     */
+    String __SessionPathProperty = "org.eclipse.jetty.servlet.SessionPath";
```
Comment (sbordet) [inline jetty-core/jetty-session/src/main/java/org/eclipse/jetty/session/SessionConfig.java]: This is core, should not have `servlet` in the package name.

File: jetty-ee10/jetty-ee10-demos/jetty-ee10-demo-embedded/src/main/resources/demo/webdefault-ee10.xml [hunk 18]
```diff
@@ -44,15 +44,15 @@
...
       <param-value>127.0.0.1</param-value> 
     </context-param> 
     <context-param>
-      <param-name>org.eclipse.jetty.ee10.servlet.SessionPath</param-name>
+      <param-name>org.eclipse.jetty.servlet.SessionPath</param-name>
       <param-value>/</param-value>
     </context-param>
     <context-param>
-      <param-name>org.eclipse.jetty.ee10.servlet.MaxAge</param-name>
+      <param-name>org.eclipse.jetty.servlet.MaxAge</param-name>
```
Comment (sbordet) [inline jetty-ee10/jetty-ee10-demos/jetty-ee10-demo-embedded/src/main/resources/demo/webdefault-ee10.xml]: No `eeN` in the package? Or `servlet` should be replaced with `session`?

File: jetty-ee10/jetty-ee10-osgi/test-jetty-ee10-osgi/src/test/config/etc/webdefault-ee10.xml [hunk 19]
```diff
@@ -52,15 +52,15 @@
...
       <param-value>127.0.0.1</param-value> 
     </context-param> 
     <context-param>
-      <param-name>org.eclipse.jetty.ee10.servlet.SessionPath</param-name>
+      <param-name>org.eclipse.jetty.servlet.SessionPath</param-name>
       <param-value>/</param-value>
     </context-param>
     <context-param>
-      <param-name>org.eclipse.jetty.ee10.servlet.MaxAge</param-name>
+      <param-name>org.eclipse.jetty.servlet.MaxAge</param-name>
```
Comment (sbordet) [inline jetty-ee10/jetty-ee10-osgi/test-jetty-ee10-osgi/src/test/config/etc/webdefault-ee10.xml]: Ditto

File: jetty-ee10/jetty-ee10-tests/jetty-ee10-test-integration/src/test/resources/webdefault-ee10.xml [hunk 20]
```diff
@@ -44,15 +44,15 @@
...
       <param-value>127.0.0.1</param-value> 
     </context-param> 
     <context-param>
-      <param-name>org.eclipse.jetty.ee10.servlet.SessionPath</param-name>
+      <param-name>org.eclipse.jetty.servlet.SessionPath</param-name>
       <param-value>/</param-value>
     </context-param>
     <context-param>
-      <param-name>org.eclipse.jetty.ee10.servlet.MaxAge</param-name>
+      <param-name>org.eclipse.jetty.servlet.MaxAge</param-name>
```
Comment (sbordet) [inline jetty-ee10/jetty-ee10-tests/jetty-ee10-test-integration/src/test/resources/webdefault-ee10.xml]: Ditto

File: jetty-ee9/jetty-ee9-demos/jetty-ee9-demo-embedded/src/main/resources/demo/webdefault-ee9.xml [hunk 21]
```diff
@@ -52,15 +52,15 @@
...
       <param-value>127.0.0.1</param-value> 
     </context-param> 
     <context-param>
-      <param-name>org.eclipse.jetty.ee9.servlet.SessionPath</param-name>
+      <param-name>org.eclipse.jetty.servlet.SessionPath</param-name>
       <param-value>/</param-value>
     </context-param>
     <context-param>
-      <param-name>org.eclipse.jetty.ee9.servlet.MaxAge</param-name>
+      <param-name>org.eclipse.jetty.servlet.MaxAge</param-name>
```
Comment (sbordet) [inline jetty-ee9/jetty-ee9-demos/jetty-ee9-demo-embedded/src/main/resources/demo/webdefault-ee9.xml]: Ditto

File: jetty-ee9/jetty-ee9-osgi/test-jetty-ee9-osgi/src/test/config/etc/webdefault-ee9.xml [hunk 22]
```diff
@@ -52,15 +52,15 @@
...
       <param-value>127.0.0.1</param-value> 
     </context-param> 
     <context-param>
-      <param-name>org.eclipse.jetty.ee9.servlet.SessionPath</param-name>
+      <param-name>org.eclipse.jetty.servlet.SessionPath</param-name>
       <param-value>/</param-value>
     </context-param>
     <context-param>
-      <param-name>org.eclipse.jetty.ee9.servlet.MaxAge</param-name>
+      <param-name>org.eclipse.jetty.servlet.MaxAge</param-name>
```
Comment (sbordet) [inline jetty-ee9/jetty-ee9-osgi/test-jetty-ee9-osgi/src/test/config/etc/webdefault-ee9.xml]: Ditto

File: jetty-ee9/jetty-ee9-tests/jetty-ee9-test-integration/src/test/resources/webdefault-ee9.xml [hunk 23]
```diff
@@ -52,15 +52,15 @@
...
       <param-value>127.0.0.1</param-value> 
     </context-param> 
     <context-param>
-      <param-name>org.eclipse.jetty.ee9.servlet.SessionPath</param-name>
+      <param-name>org.eclipse.jetty.servlet.SessionPath</param-name>
       <param-value>/</param-value>
     </context-param>
     <context-param>
-      <param-name>org.eclipse.jetty.ee9.servlet.MaxAge</param-name>
+      <param-name>org.eclipse.jetty.servlet.MaxAge</param-name>
```
Comment (sbordet) [inline jetty-ee9/jetty-ee9-tests/jetty-ee9-test-integration/src/test/resources/webdefault-ee9.xml]: Ditto

File: jetty-ee9/jetty-ee9-webapp/src/main/config/etc/webdefault-ee9.xml [hunk 24]
```diff
@@ -44,15 +44,15 @@
...
       <param-value>127.0.0.1</param-value> 
     </context-param> 
     <context-param>
-      <param-name>org.eclipse.jetty.ee9.servlet.SessionPath</param-name>
+      <param-name>org.eclipse.jetty.servlet.SessionPath</param-name>
       <param-value>/</param-value>
     </context-param>
     <context-param>
-      <param-name>org.eclipse.jetty.ee9.servlet.MaxAge</param-name>
+      <param-name>org.eclipse.jetty.servlet.MaxAge</param-name>
```
Comment (sbordet) [inline jetty-ee9/jetty-ee9-webapp/src/main/config/etc/webdefault-ee9.xml]: Ditto

**Review by sbordet — CHANGES_REQUESTED** [review]

File: documentation/jetty-documentation/src/main/asciidoc/programming-guide/server/sessions/session-sessioncache.adoc (hunk 8 above)
Comment (janbartel) [reply documentation/jetty-documentation/src/main/asciidoc/programming-guide/server/sessions/session-sessioncache.adoc]: I've modified the wording a little differently.

File: documentation/jetty-documentation/src/main/asciidoc/programming-guide/server/sessions/session-sessioncache.adoc (hunk 10 above)
Comment (janbartel) [reply documentation/jetty-documentation/src/main/asciidoc/programming-guide/server/sessions/session-sessioncache.adoc]: I've modified the wording a little differently.

File: documentation/jetty-documentation/src/main/asciidoc/programming-guide/server/sessions/session-sessionhandler.adoc (hunk 11 above)
Comment (janbartel) [reply documentation/jetty-documentation/src/main/asciidoc/programming-guide/server/sessions/session-sessionhandler.adoc]: That text was from an older version and no longer exists.

File: jetty-core/jetty-deploy/src/test/resources/etc/webdefault.xml (hunk 15 above)
Comment (janbartel) [reply jetty-core/jetty-deploy/src/test/resources/etc/webdefault.xml]: No, all session code except for `SessionHandler` is environment agnostic, and thus the config params.

**Pull request merged**